rllib/env/remote_base_env.py

```diff
@@ -49,71 +55,88 @@ def __init__(self,
             existing_envs: Optional list of already created sub-environments.
                 These will be used as-is and only as many new sub-envs as
                 necessary (`num_envs - len(existing_envs)`) will be created.
+            worker: An optional RolloutWorker that owns the env. This is only
+                used if `remote_worker_envs` is True in your config and the
+                `on_sub_environment_created` custom callback needs to be
+                called on each created actor.
         """
 
         # Could be creating local or remote envs.
         self.make_env = make_env
-        # Whether the given `make_env` callable already returns ray.remote
-        # objects or not.
-        self.make_env_creates_actors = False
-        # Already existing env objects (generated by the RolloutWorker).
-        self.existing_envs = existing_envs or []
         self.num_envs = num_envs
         self.multiagent = multiagent
         self.poll_timeout = remote_env_batch_wait_ms / 1000
+        # Already existing env objects (generated by the RolloutWorker).
+        self.existing_envs = existing_envs or []
+        self.worker = worker
+
+        # Whether the given `make_env` callable already returns ray.remote
+        # objects or not.
+        self.make_env_creates_actors = False
+
+        self._observation_space = None
+        self._action_space = None
 
         # List of ray actor handles (each handle points to one @ray.remote
         # sub-environment).
         self.actors: Optional[List[ray.actor.ActorHandle]] = None
-        self._observation_space = None
-        self._action_space = None
+
+        # `self.make_env` already produces Actors: Use it directly.
+        if len(self.existing_envs) > 0 and isinstance(self.existing_envs[0],
+                                                      ray.actor.ActorHandle):
+            self.make_env_creates_actors = True
+            self.actors = []
+            while len(self.actors) < self.num_envs:
+                sub_env = self.make_env(len(self.actors))
```

Comment:
I know you are just moving code, but I am really confused here.
self.actors is empty at first, so at the start, we will do
sub_env = self.make_env(0), which I hope is producing 0 envs. :)
then this becomes an infinite loop??

Reply:
Yeah, moved some stuff :) Just don't like this lazy creation of properties in the `poll()` method. Stuff like this should always be in the `__init__()/setup()` methods. Let me see ...

Reply:
Yup, something wrong here. I fixed it:

```
# Initially:
self.actors = existing_envs  # <- forgot this as we may already have n sub-envs created beforehand

# same in the following else block:
            self.actors = existing_envs + [   # <- existing envs must be prepended
                make_remote_env(i) for i in range(self.num_envs - len(existing_envs))
            ]
```